### DISCUSSION TITLE
Clarify concerns about UMA Token endpoint performance

### DISCUSSION THREAD

**Discussion by CarloDeSutter** [keycloak/keycloak#25821] [Q&A]
Hi,

I am not directly creating an issue because we are not yet working on the latest server version.

We are, for now, looking for some clarification of the following situation, as we are wondering if we are doing something that is not Keycloak-idiomatic in terms of calling the token endpoint for RPTs, and we need to dive in further at our end. In other words, I wondering if this would be fixed by simply upgrading and we might have to look into other factors at play here.

We are having some very strange behaviour when it comes to requesting RPT's through the token endpoint, grant type urn:ietf:params:oauth:grant-type:uma-ticket.

**How to Reproduce?**
- Create confidential client **foo** with Authorization
- Add a few hundreds (it will be noticeable as soon as 100 or worse) Authorization resources in client **foo**, with one or two scopes
- Allow access to all of them for your user
- Call the /token endpoint  with grant_type urn:ietf:params:oauth:grant-type:uma-ticket, audience **foo**
- At this point we get back a response containing all those resources with their scopes

**Expected behaviour**
Doing **only** a few calls at the same time do not consistently take over 2000ms.

**Actual behaviour**
We are seeing really strange behaviour, in that if you make a few calls at the same time to this endpoint, we get 100-200ms tops.
As soon as more calls are done at the same time (there is a bit more load) we are seeing consistent increase of endpoint RTT up to 20000+ms. Rebooting the Pods seems to temporarily help, same goes with a clear Realm and User cache in v12. (no longer available in v21).

Request and response payload between "normal" and "slow"calls is no different.

**Extra info**
We have:
- Staging which was recently upgraded from v12 to v21 without issues but with a lot less load and resources are getting degraded performance (from 100ms in v12 to over 2000ms for one RPT call with little to no resources), this is consistent after the upgrade.
- Production sitting on v12, not yet upgraded because Staging was only upgraded very recently and we are seeing degraded performance. Production suffers from the bevaviour described above.

We are using a Helm chart setup, the Pods have sufficient RAM and CPU and are hosted in western Europe. The MySQL DB sits in AWS. At first sight this seems DB(indirectly via network or directly)-bound, but I would expect that this is some weird caching bug (as we have some means to improve it very shortly) and that this would hopefully be solved by upgrading.

Does this ring a bell to anyone?

I am having a hard time finding info on what our expectations of getting RPT tokens should be and how it scales.

**Comment by CarloDeSutter** [comment]
Friendly bump.

**Comment by CarloDeSutter** [comment]
Bit more info from our end

Staging (KC v21): 
- Helm chart: https://github.com/codecentric/helm-charts/tree/master/charts/keycloakx (2.2.2)
- AWS managed mySQL instance replaced from older mySQL to a v8.x

Production (KC v12): 
- Helm chart https://github.com/codecentric/helm-charts/tree/master/charts/keycloak (10.2.0)
- AWS managed mySQL v5.x

I talked about how clearing cache helps on the v12 instance, notice the effect of (temporary improves) a realm cache clear I did yesterday:

![image](https://github.com/keycloak/keycloak/assets/20969135/0146f8b1-4264-4471-8a17-c717d40c19e6)

A common denominator that we are noticing is that it tends to worsen when using a shared (AWS) mysql instance where more of other stuff is using it. Our infrastructure team did however some checks on health, load and so on but so far haven't noticed anything that explains this behaviour. We are planning on doing more analysing on this though.

Is there some information or experience from your side that could confirm we should, for example, get a full dedicated mysql, and/or better specs? I am having trouble finding what type of machine this should be, apologies if I am missing some documentation.

Thanks for chiming in, this would be of great help.

**Reply by sschu** [reply]
Do you have some tracing agent running? This would allow you to see the sql queries emitted by a certain request and see how long they take...

**Reply by CarloDeSutter** [reply]
Thanks for your reply @sschu 

We will be diving deeper at the database level asap, what does stick out however for example for the v12 version:
![image](https://github.com/keycloak/keycloak/assets/20969135/40363763-c631-4c2c-89bc-e0c2dd3ce0fe)
Notice the drop as soon as I hit the Realm cache clear button, this is temporary after a few days things start to ramp up again.
The load by users is quite level every working day.

The same effect can be observed by rebooting the pod.

Is there any trace of this being a major bug that is solved in some version later, because I am unable to pinpoint it.

For the v12>v21 version update what comes out is the following increase of load just after switching to the v21 instance (which uses another mysql instance, v8.x)

v12 before update:

![image](https://github.com/keycloak/keycloak/assets/20969135/70ace7a3-3559-4f23-a2f6-460cfacb8a27)

after update:
![image](https://github.com/keycloak/keycloak/assets/20969135/eab5eae6-6b27-4101-9d7a-4957adb1ed84)

I have more metrics but the gist I want to present here is that for some reason the load increased by updating:
- v12 to v21
- mysql v5 to v8

**Reply by CarloDeSutter** [reply]
To be clear, the response times vary greatly while the request is exactly the same (frequency can vary a bit of course, but we are not talking huge numbers here).

**Reply by sschu** [reply]
I think there is no way around finding out where exactly the time is spent i.e. use an agent to see if time is spent in code or in db and if in db, check the queries that are executed.

**Reply by CarloDeSutter** [reply]
Hi as an example **(this used to take 100ms in v12)**. now over 2s:
- client has about 90 resources, and for each resource we set up a policy and permission

~ 100x "SELECT keycloak-staging-a.RESOURCE_SERVER_POLICY"
~ 100x "SELECT keycloak-staging-a.POLICY_CONFIG"
~ 100x just "SELECT keycloak-staging-a" 

This results in about 300+ SQL statements, and scales with the number of resources we need to add and protect.

**Reply by CarloDeSutter** [reply]
I guess my question is, is this normal that this requires so many SQL statements to be executed?

If you feel this is a non-technical issue, what is the idiomatic/correct way to authorize many (many is relative here) resources without the token endpoint becoming a bottleneck?

**Reply by sschu** [reply]
Did the number of queries increase when going from v12 to v23? Or is the number the same but the request processing became slower?

**Reply by CarloDeSutter** [reply]
Hi @sschu after doing some searching again, we might try upgrading staging from v21 to v22 as this might be a factor for that environment: https://github.com/keycloak/keycloak/pull/20566

**Reply by sschu** [reply]
@CarloDeSutter Yeah, that issue looks promising.

**Reply by ptschmidt** [reply]
Hi, we're facing the same issue as @CarloDeSutter. We upgraded from v12 to v20.0.5 and observed a massive increase in uma-ticket responses. In our case it's possible that a single user has access to hundreds of resources of a client. Since we changed a lot of stuff and upgraded not only the KC version but the database (single node mariadb to galera) and some infrastructure as well, we checked database and the other changed components first. After reading @CarloDeSutter's yestersdays comment regarding #20566 we upgraded from v20.0.5 to v22.0.5. Our first tests are looking promising! I'm interested in your experiences and test results and I'm looking forward to hear about them @CarloDeSutter

**Reply by CarloDeSutter** [reply]
Thanks for chiming in @ehwhw0w.

I can confirm the same observations.

We have upgraded from v21 to v22 yesterday and are observing dramatic improvements when it comes to POST /token with grant_type urn:ietf:params:oauth:grant-type:uma-ticket.

This brings some confidence back that upgrading our Production setup from v12 to v22 should also fix the performance issues with the same API call over there.
